### PR TITLE
[PLUGIN-731] Added support for BIGNUMERIC for BigQuery Source and Sink (Type is ch…

### PR DESCRIPTION
**Pull request by sechegaray**
…osen depending on the precision and scale of the Decimal field

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java [hunk 1]
```diff
@@ -263,6 +263,10 @@ public static Schema convertFieldType(Field field, @Nullable FailureCollector co
         // bigquery has 38 digits of precision and 9 digits of scale for NUMERIC.
         // https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types
         return Schema.decimalOf(38, 9);
+      case BIGNUMERIC:
+        // bigquery has 76 digits of precision and 38 digits of scale for BIGNUMERIC.
+        // https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types
+        return Schema.decimalOf(77, 38);
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java]: The comment says 76 and code 77.

**Review by tivv — COMMENTED** [review]
Also please add some tests

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java (hunk 1 above)
Comment (tivv) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java]: BTW: With the number of places this magic numbers are used, it makes sense to introduce constants for them

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryTypeSize.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * the License.
+ */
+
+package io.cdap.plugin.gcp.bigquery.util;
+
+/**
+ * Stores the Precision and Scale for the different Numeric (Decimal) types
+ * provided by BigQuery
+ */
+public final class BigQueryTypeSize {
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryTypeSize.java:24]: I would add a link here to the BQ docs where this specification is outlined, just for future reference

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java [hunk 3]
```diff
@@ -333,12 +339,12 @@ public static ValidationFailure validateFieldSchemaMatches(Field bqField, Schema
...
 
-      // BigQuery schema precision must be at most 38 and scale at most 9
+      // BigQuery schema precision must be at most 77 and scale at most 38
       if (logicalType == Schema.LogicalType.DECIMAL) {
-        if (fieldSchema.getPrecision() > 38 || fieldSchema.getScale() > 9) {
+        if (fieldSchema.getPrecision() > BigNumeric.PRECISION || fieldSchema.getScale() > BigNumeric.SCALE) {
           return collector.addFailure(String.format("Decimal Field '%s' has invalid precision '%s' and scale '%s'. ",
                                                     name, fieldSchema.getPrecision(), fieldSchema.getScale()),
-                                      "Precision must be at most 38 and scale must be at most 9.");
+                                      "Precision must be at most 77 and scale must be at most 38.");
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java]: nit: the values in these comment should also come from the static values you defined.

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryTypeSize.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * Precision and scale of BigQuery Numeric class
+   */
+  public static final class Numeric {
+    public static final int PRECISION = 38;
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryTypeSize.java:30]: Make it MAX_PRECISION and similarly others.

Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryTypeSize.java:30]: resolving after our conversation earlier today.

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java (hunk 3 above)
Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java]: fixed

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryTypeSize.java (hunk 2 above)
Comment (sechegaray) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryTypeSize.java:24]: fixed

**Review by tivv — APPROVED** [review]

**Pull request merged**